### PR TITLE
Fixed #22423 - Added support for MySQL operators on real geometries

### PR DESCRIPTION
**Pull request by claudep**

### CODE REVIEW

File: docs/ref/contrib/gis/geoquerysets.txt [hunk 1]
```diff
@@ -148,10 +148,15 @@ Backend     SQL Equivalent
...
 PostGIS     ``ST_Contains(poly, geom)``
 Oracle      ``SDO_CONTAINS(poly, geom)``
-MySQL       ``MBRContains(poly, geom)``
+MySQL       ``ST_Contains(poly, geom)``
 SpatiaLite  ``Contains(poly, geom)``
 ==========  ============================
 
+.. versionchanged:: 3.0
+
+    MySQL now uses ``ST_Contains`` and operates on real geometries instead of on
```
Comment (pope1ni) [inline docs/ref/contrib/gis/geoquerysets.txt]: Do we want to say what function was used before the change in all of these cases?

e.g. `MySQL now uses ``ST_Contains`` instead of ``MBRContains`` and ...`

(This would give some context while the `.. versionchanged::` annotation is in place.)

File: docs/ref/contrib/gis/geoquerysets.txt [hunk 2]
```diff
@@ -453,11 +494,16 @@ Example::
 Backend     SQL Equivalent
 ==========  ==========================
 PostGIS     ``ST_Touches(poly, geom)``
-MySQL       ``MBRTouches(poly, geom)``
+MySQL       ``ST_Touches(poly, geom)``
```
Comment (pope1ni) [inline docs/ref/contrib/gis/geoquerysets.txt]: Probably one for a separate commit, but these tables are sometimes ordered...
`PostGIS → Oracle → MySQL → SpatiaLite`
...and sometimes...
`PostGIS → MySQL → Oracle → SpatiaLite`

File: docs/ref/contrib/gis/geoquerysets.txt [hunk 3]
```diff
@@ -372,6 +397,22 @@ Oracle                      ``SDO_GEOM.VALIDATE_GEOMETRY_WITH_CONTEXT(poly, 0.05
...
+Tests if the geometry field spatially overlaps the lookup geometry.
+
+==========  =================================================
+Backend     SQL Equivalent
+==========  =================================================
+PostGIS     ``ST_Overlaps(poly, geom)``
+Oracle      ``SDO_OVERLAPS(poly, geom)``
+MySQL       ``ST_Overlaps(poly, geom)``
+SpatiaLite  ``Overlaps(poly, geom)``
+==========  =================================================
```
Comment (pope1ni) [inline docs/ref/contrib/gis/geoquerysets.txt]: If this table was missing, should it be added as a separate commit first to keep the diff to a minimum?

File: docs/releases/3.0.txt [hunk 4]
```diff
@@ -64,7 +64,8 @@ Minor features
 :mod:`django.contrib.gis`
 ~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ...
+* Added support on MySQL for spatial lookup functions operating on real
+  geometries instead of mere bounding boxes.
```
Comment (pope1ni) [inline docs/releases/3.0.txt]: Drop "mere".

File: tests/gis_tests/geoapp/tests.py [hunk 5]
```diff
@@ -221,9 +222,22 @@ def test_empty_geometries(self):
             self.assertEqual(feature.geom.srid, g.srid)
 
 
+def mysql_disjoint_bug(func):
+    """
+    It's still unknown why the MySQL disjoint lookup gives different results
```
Comment (pope1ni) [inline tests/gis_tests/geoapp/tests.py]: How different is this? Is it something that should be documented?

File: docs/ref/contrib/gis/db-api.txt [hunk 6]
```diff
@@ -25,21 +25,12 @@ GeoDjango currently provides the following spatial database backends:
...
-    according to the specification.  Those that are implemented return
-    the same result as the corresponding MBR-based functions.
-
-In other words, while spatial lookups such as :lookup:`contains <gis-contains>`
-are available in GeoDjango when using MySQL, the results returned are really
-equivalent to what would be returned when using :lookup:`bbcontains`
-on a different spatial backend.
+Before MySQL 5.6.1, spatial extensions only supported bounding box operations
+(what MySQL calls minimum bounding rectangles, or MBR). Specifically, MySQL did
+not conform to the OGC standard. Django 3.0 introduced support for spatial
```
Comment (timgraham) [inline docs/ref/contrib/gis/db-api.txt]: I would move the "Django 3.0" mention to a versionchanged annotation if you feel it's required so as not require retranslations when it's later obsolete.

File: docs/ref/contrib/gis/geoquerysets.txt [hunk 7]
```diff
@@ -148,10 +148,15 @@ Backend     SQL Equivalent
...
 PostGIS     ``ST_Contains(poly, geom)``
 Oracle      ``SDO_CONTAINS(poly, geom)``
-MySQL       ``MBRContains(poly, geom)``
+MySQL       ``ST_Contains(poly, geom)``
 SpatiaLite  ``Contains(poly, geom)``
 ==========  ============================
 
+.. versionchanged:: 3.0
+
+    MySQL now uses ``ST_Contains`` instead of ``MBRContains`` and operates on
```
Comment (timgraham) [inline docs/ref/contrib/gis/geoquerysets.txt]: The usual wording is something like, "In older versions, MySQL uses MBRContains instead of ST_Contains and operates only on bounding boxes."

File: docs/ref/contrib/gis/geoquerysets.txt [hunk 8]
```diff
@@ -303,9 +318,9 @@ SpatiaLite  ``Equals(poly, geom)``
 *Availability*: `PostGIS <https://postgis.net/docs/ST_Geometry_Same.html>`__,
 Oracle, MySQL, SpatiaLite, PGRaster (Bilateral)
 
-Tests if the geometry field is "equal" to the lookup geometry. On Oracle and
-SpatiaLite it tests spatial equality, while on MySQL and PostGIS it tests
-equality of bounding boxes.
+Tests if the geometry field is "equal" to the lookup geometry. On Oracle, MySQL
```
Comment (timgraham) [inline docs/ref/contrib/gis/geoquerysets.txt]: MySQL,

File: docs/ref/contrib/gis/geoquerysets.txt [hunk 9]
```diff
@@ -303,9 +318,9 @@ SpatiaLite  ``Equals(poly, geom)``
...
+and SpatiaLite it tests spatial equality, while on PostGIS it tests equality of
```
Comment (timgraham) [inline docs/ref/contrib/gis/geoquerysets.txt]: SpatiaLite,

File: docs/ref/contrib/gis/geoquerysets.txt (hunk 3 above)
Comment (timgraham) [reply docs/ref/contrib/gis/geoquerysets.txt]: Yes, since that change could be backported.

File: docs/releases/3.0.txt [hunk 10]
```diff
@@ -64,7 +64,8 @@ Minor features
 :mod:`django.contrib.gis`
 ~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ...
+* Added support on MySQL for spatial lookup functions operating on real
```
Comment (timgraham) [inline docs/releases/3.0.txt]: Allowed MySQL spatial lookup functions to operate on real geometries. Previous support was limited to bounding boxes.

File: tests/gis_tests/geoapp/tests.py (hunk 5 above)
Comment (timgraham) [reply tests/gis_tests/geoapp/tests.py]: chop "still" and "in our tests"

**Comment by timgraham** [comment]
Uses two dashes in commit message.

**Comment by tuky** [comment]
if you cannot wait for django 3.0, you can backport this locally by doing

```python
import django
from django.contrib.gis.db.backends.utils import SpatialOperator
from django.contrib.gis.db.models.fields import BaseSpatialField
from django.contrib.gis.db.models.lookups import GISLookup      


assert django.VERSION < (3,)  # django 3 ships this OOB, so remove custom GISLookup on upgrading


@BaseSpatialField.register_lookup                   
class ContainsLookup(GISLookup):                    
    lookup_name = 'contains'                        
                                                    
    def get_rhs_op(self, connection, rhs):          
        return SpatialOperator(func='ST_Contains')  
```
